packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewFlutterPlugin.java

```diff
@@ -38,36 +53,113 @@ public WebViewFlutterPlugin() {}
+    WebChromeClientHostApi.setup(
+        binaryMessenger,
+        new WebChromeClientHostApiImpl(
+            instanceManager,
+            new WebChromeClientHostApiImpl.WebChromeClientCreator(),
+            new WebChromeClientFlutterApiImpl(binaryMessenger, instanceManager)));
+    DownloadListenerHostApi.setup(
+        binaryMessenger,
+        new DownloadListenerHostApiImpl(
+            instanceManager,
+            new DownloadListenerHostApiImpl.DownloadListenerCreator(),
+            new DownloadListenerFlutterApiImpl(binaryMessenger, instanceManager)));
+    WebSettingsHostApi.setup(
+        binaryMessenger,
+        new WebSettingsHostApiImpl(
+            instanceManager, new WebSettingsHostApiImpl.WebSettingsCreator()));
+  }
+
   @Override
-  public void onAttachedToEngine(FlutterPluginBinding binding) {
-    BinaryMessenger messenger = binding.getBinaryMessenger();
-    binding
-        .getPlatformViewRegistry()
-        .registerViewFactory(
-            "plugins.flutter.io/webview",
-            new FlutterWebViewFactory(messenger, /*containerView=*/ null));
-    flutterCookieManager = new FlutterCookieManager(messenger);
+  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
+    pluginBinding = binding;
+    setUp(
+        binding.getBinaryMessenger(),
+        binding.getPlatformViewRegistry(),
+        binding.getApplicationContext(),
+        null);
   }
 
   @Override
-  public void onDetachedFromEngine(FlutterPluginBinding binding) {
+  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
     if (flutterCookieManager == null) {
       return;
     }
 
     flutterCookieManager.dispose();
     flutterCookieManager = null;
   }
+
+  @Override
+  public void onAttachedToActivity(@NonNull ActivityPluginBinding activityPluginBinding) {
+    final Activity activity = activityPluginBinding.getActivity();
+    webViewHostApi.setContext(activityPluginBinding.getActivity());
+    javaScriptChannelHostApi.setPlatformThreadHandler(new Handler(activity.getMainLooper()));
+  }
+
+  @Override
+  public void onDetachedFromActivityForConfigChanges() {
+    final Context context = pluginBinding.getApplicationContext();
+    webViewHostApi.setContext(context);
+    javaScriptChannelHostApi.setPlatformThreadHandler(new Handler(context.getMainLooper()));
+  }
+
+  @Override
+  public void onReattachedToActivityForConfigChanges(
+      @NonNull ActivityPluginBinding activityPluginBinding) {
+    final Activity activity = activityPluginBinding.getActivity();
+    webViewHostApi.setContext(activity);
+    javaScriptChannelHostApi.setPlatformThreadHandler(new Handler(activity.getMainLooper()));
+  }
+
+  @Override
+  public void onDetachedFromActivity() {
+    final Context context = pluginBinding.getApplicationContext();
+    webViewHostApi.setContext(context);
+    javaScriptChannelHostApi.setPlatformThreadHandler(new Handler(context.getMainLooper()));
```

Comment:
This pattern is repeated four times; it seems like changing this (e.g., adding another thing that needs to be adjusted) would be very error-prone. How about extracting a helper that takes the `Context`, and then each of these can be a one-liner to call that helper with the right context for the scenario.